model/riscv_pmp_control.sail

```diff
@@ -104,15 +104,46 @@ function pmpCheckRWX(ent, acc) = {
   }
 }
 
-// this needs to be called with the effective current privilege.
-val pmpCheckPerms: (Pmpcfg_ent, AccessType(ext_access_type), Privilege) -> bool
+/* this needs to be called with the effective current privilege */
+val pmpCheckPerms: (Pmpcfg_ent, AccessType(ext_access_type), Privilege) -> bool effect {rreg}
 function pmpCheckPerms(ent, acc, priv) = {
-  match priv {
-    Machine => if   pmpLocked(ent)
-               then pmpCheckRWX(ent, acc)
-               else true,
-    _       => pmpCheckRWX(ent, acc)
+  if mseccfg.MML() == 0b1 then {
+    if ent.R() == 0b0 & ent.W() == 0b1 then {
+      if pmpLockBit(ent) then {
+        match acc {
+          Execute(_) => true,
+          Read(_)    => priv == Machine & ent.X() == 0b1,
+           _         => false
+        }
+      }
+      else {
+        match acc {
+          Read(_)   => true,
+          Write(_)  => priv == Machine | ent.X() == 0b1,
+            _       => false
+        }
+      }
+    }
+    else {
+      if ent.R() == 0b1 & ent.W() == 0b1 & ent.X() == 0b1 & pmpLockBit(ent) then {
+        match acc {
+          Read(_) => true,
+            _     => false
+        }
+      }
+    else {
+      if (priv == Machine) == pmpLockBit(ent) then 
+        pmpCheckRWX(ent, acc)
+      else false
+    }}
   }
+  else 
+    match priv {
+      Machine => if pmpLockBit(ent) then 
+                 pmpCheckRWX(ent, acc)
+                 else true,
+        _     => pmpCheckRWX(ent, acc)
+    }
```

Comment:
You have trailing whitespace on some of these lines

Reply:
Fixed